### DISCUSSION TITLE
Is Asami like Datomic?

### DISCUSSION THREAD

**Discussion by kyleerhabor** [threatgrid/asami#210] [Q&A] [answered]
I've been evaluating my options to use for a database in my project. Datomic has a lot of appeals, but since it's proprietary and coupled with AWS, I've sought other options. Eva is interesting, though it seems to be unmaintained (though it has all the concepts I like). Datahike seems very similar to Asami.

One of the core principles I like about Datomic is data is never deleted (unless absolutely necessary). Does Asami manage data like this as well?

**Comment by simongray** [comment]
Yes. Reading through the wiki makes it clear. You have the same `as-of` and other time travel functions of the Datomic API as with Datomic proper.

**Comment by lilactown** [comment, marked as answer]
A point of clarity:
Datomic Cloud is a beast in and of itself, very different from both Datomic On-Prem and Asami. It handles not only typical "database" stuff but also code deployments, auto scaling, provisioning multiple clusters, and running your code in a "Function-as-a-Service" kind of way. It is AWS only, which you pointed out is a deal breaker for you, but just saying that Datomic Cloud is A LOT of stuff than just a database.

It sounds like, since you don't want to use AWS, that you could use Datomic OnPrem and handle provisioning and deploying the Datomic database yourself on your hosting provider of choice. In which case, I think you'll find that there are still quite a few differences between it and Asami:

1. Datomic has a distributed architecture that allows it to scale reads horizontally. Asami doesn't have anything out of the box that supports distribution; its API assumes that queries are executed on the same machine that has the backing store (correct me if I'm wrong)
2. Datomic already has storage backends for many different durable storage services like sql databases, cassandra, s3, etc. As of now, asami has only in-memory and file backends AFAICT. You can of course implement your own
3. Datomic has been used in production at scale in many different environments and has paid support. I haven't seen nearly as many people discussing using Asami as their production database, and AFAIK it does not have built out corporate support service. Asami is free and open source
4. Datomic requires a schema, Asami is open world
5. Datomic only allows data to be loaded as EDN, i.e. attributes must be keywords and values must be scalar or references to other entities. Asami allows attributes and values to be any type, even complex ones.
6. Datomic does not do any optimizing of your query. Asami has a query planner which can optimize your query for you. Sometimes you have to turn it off, though.
7. Datomic only runs on the JVM. Asami can run on the JVM and JS environments (via CLJS).

There are probably numerous other smaller differences between the two, which might be more or less relevant to you depending on what kind of application you're trying to build. I laid out the ones above that seem relevant to me and I would consider before picking one or the other.

There are many cases where I might be building an application that I know isn't going to scale past one service/node, isn't going to need paid support and where Asami's open world assumption could be beneficial, i.e. a personal finance app that I run locally on my machine or put on a small VPS somewhere, or an in-memory store for a microservice.

If however I was looking for a solution that would serve as the primary system of record for a large distributed system, I would pick datomic right now.

Hope that helps, and please correct me if I'm wrong about any of this. I would also love to hear how threatgrid uses it, as that may help me understand this question better as well!
